src/main/kotlin/com/numberone/daepiro/domain/auth/service/AuthService.kt

```diff
@@ -0,0 +1,84 @@
+import com.numberone.daepiro.global.feign.KakaoFeign
+import com.numberone.daepiro.global.feign.NaverFeign
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.security.crypto.password.PasswordEncoder
+import org.springframework.stereotype.Service
+
+@Service
+class AuthService(
+    private val userRepository: UserRepository,
+    private val passwordEncoder: PasswordEncoder,
+    private val kakaoFeign: KakaoFeign,
+    private val naverFeign: NaverFeign,
+    @Value("\${jwt.secret-key}") private val secretKey: String,
+    @Value("\${jwt.access-token-expire}") private val accessTokenExpire: Long,
+    @Value("\${jwt.refresh-token-expire}") private val refreshTokenExpire: Long,
+    @Value("\${jwt.admin-token-expire}") private val adminTokenExpire: Long
+) {
+    fun kakaoLogin(
+        request: SocialLoginRequest
+    ): ApiResult<TokenResponse> {
+        val userInfo = kakaoFeign.getUserInfo(JwtUtils.PREFIX_BEARER + request.socialToken)
+            ?: throw CustomException(INVALID_SOCIAL_TOKEN)
+        val socialId = userInfo.id.toString()
+        val user = userRepository.findBySocialLoginInformation_SocialId(socialId)
+            ?: userRepository.save(UserEntity.of(SocialPlatform.KAKAO, socialId))
+
+        return ApiResult.ok(TokenResponse.of(user, secretKey, accessTokenExpire, refreshTokenExpire))
+    }
+
+    fun naverLogin(
+        request: SocialLoginRequest
+    ): ApiResult<TokenResponse> {
+        val userInfo = naverFeign.getUserInfo(JwtUtils.PREFIX_BEARER + request.socialToken)
+            ?: throw CustomException(INVALID_SOCIAL_TOKEN)
+        val socialId = userInfo.response.id
+        val user = userRepository.findBySocialLoginInformation_SocialId(socialId)
+            ?: userRepository.save(UserEntity.of(SocialPlatform.NAVER, socialId))
+
+        return ApiResult.ok(TokenResponse.of(user, secretKey, accessTokenExpire, refreshTokenExpire))
+    }
+
+    fun adminLogin(
+        request: AdminLoginRequest
+    ): ApiResult<TokenResponse> {
+        val user = userRepository.findByPasswordLoginInformation_Username(request.username)
+            ?: throw CustomException(NOT_FOUND_USER)
+        require(user.passwordLoginInformation != null)
+        if (!passwordEncoder.matches(request.password, user.passwordLoginInformation!!.password))
+            throw CustomException(INVALID_PASSWORD)
+
+        return ApiResult.ok(TokenResponse.of(user, secretKey, adminTokenExpire, adminTokenExpire))
+    }
+
+    fun refreshToken(
+        request: RefreshTokenRequest
+    ): ApiResult<TokenResponse> {
+        // todo redis에서 refresh token 관리하도록 변경
+        val tokenInfo = JwtUtils.extractInfoFromToken(request.refreshToken, secretKey)
+        if (tokenInfo.type != TokenType.REFRESH)
+            throw CustomException(INVALID_TOKEN)
+        val user = userRepository.findByIdOrNull(tokenInfo.id)
```

Comment:
아래처럼 userRepository 에 대한 확장함수를 구현하는 것은 어때요?

```suggestion
        val user = userRepository.findByIdOrThrow(tokenInfo.id)
```

Reply:
확장함수를 구현해놓으면 서비스 로직에서 매번 예외처리를 해줄 필요가 없네요! 좋은 생각 같네요! 아래 글 참고해서 한번 만들어보겠습니다.
https://unluckyjung.github.io/jpa/kotlin/2023/03/12/JPA-kotlin-extenstion-fun/

Reply:
오 좋은 글 공유 감사드립니다!